pkg/tsdb/prometheus/time_series_query.go

```diff
@@ -397,7 +397,19 @@ func exemplarToDataFrames(response []apiv1.ExemplarQueryResult, query *Prometheu
 	// TODO: this preallocation is very naive.
 	// We should figure out a better approximation here.
 	events := make([]ExemplarEvent, 0, len(response)*2)
-
+	// Prometheus treats empty value as same as null, so `event.Labels` may not be consistent across `events`,
+	// leading errors like "frame has different field lengths, field 0 is len 5 but field 14 is len 2", need a fix.
+	eventLabels := make(map[string]struct{})
+	for _, exemplarData := range response {
+		for label := range exemplarData.SeriesLabels {
```

Comment:
I am wondering why is this also mixing in SeriesLabels. As far as I can judge, most of the time they are totally different from the exemplar labels, so I would assume most of the time they will just create additional columns that will be empty for all exemplars.

Reply:
@aocenas  Because Prometheus treats empty value as same as null. For example, I have a metrics 
`foo_metric{foo="bar"}`, the label value could be empty `foo_metric{foo=""}`, Prometheus will treat it as null, so the exemplar API may response 
```json
[
{
"seriesLabels": {
"__name__": "some_name",
"foo": "bar",
},
"exemplars": [
{
"timestamp": 1646267809.018,
"labels": {
"traceID": "xxx"
},
"value": "1"
},
{
"timestamp": 1646268870.739,
"labels": {
"traceID": "xxxxx"
},
"value": "1"
},
]
},
{
"seriesLabels": {
"__name__": "foo_metric", 
// inconsistent seriesLabels
},
"exemplars": [
{
"timestamp": 1646235461.953,
"labels": {
"traceID": "xxxx"
},
"value": "1"
}
]
},
```

Could you give me some detailed advice on code? Thank you.

Reply:
Hey, sorry for being a bit slow on this. Still though I am not sure I understand. In the example response the labels on the first series are `"__name__": "some_name", "foo": "bar"` and on the exemplars `"traceID": "xxx"`. If I understand this code, it will add all those labels to exemplars data frame. So the exemplars data frame will have columns `__name__`, `foo` and `traceID`. As there are values only for the `traceID` label only that column will have values and the `__name__` and `foo` will be just empty. If I am right I am not sure how that helps.

Reply:
Yes, the logic code is as your understand. It can helps to ensure the length is equal across all exemplar labels, fix this error raised by https://github.com/grafana/grafana-plugin-sdk-go/blob/main/data/frame.go#L278 when frame encode to JSON:

t=2022-03-02T20:52:08+0800 lvl=eror msg="Error writing to response" logger=context userId=1 orgId=1 uname=admin err="frame has different field lengths, field 0 is len 81 but field 11 is len 12"

you can see more in the issue link, thank you.

Reply:
hi @aocenas Could you take look when convenient? Thank you.

Reply:
Yes but that happens because exemplars have different labels. So it makes sense to align `exemplar.Labels` but it does not make sense to me to add `exemplarData.SeriesLabels`. `exemplarData.SeriesLabels` will just create empty columns that are not used and not needed anywhere. It seems like what will fix this is aligning just the `exemplar.Labels` and that should be enough.

Reply:
I see, I have update the code, please help review, thank you

Reply:
Hey @hanjm, sorry for this suggestion. I realized now I was wrong here and due to that this wasn't completely fixed, again sorry for that. Created a PR with similar code to yours https://github.com/grafana/grafana/pull/49924 so if you still had issues this should solve it.